cmd/migration-tool/main.go

```diff
@@ -0,0 +1,524 @@
+	flagset.AddGoFlag(original.Lookup("v"))
+
+	if err := cmd.Execute(); err != nil {
+		klog.Exitf("Execute error: %v", err)
+	}
+}
+
+func (o *options) Run() error {
+	ctx := context.Background()
+	ctxWithCancel, cancelFunction := context.WithCancel(ctx)
+
+	defer func() {
+		cancelFunction()
+	}()
+
+	stopCh := make(chan bool)
+
+	inClusterCfg, err := loadClusterConfig(&clientcmd.ConfigOverrides{})
+	imageClient, err := imageclientset.NewForConfig(inClusterCfg)
+	if err != nil {
+		return fmt.Errorf("unable to create image client: %v", err)
+	}
+
+	releasePayloadClient, err := versioned.NewForConfig(inClusterCfg)
+	if err != nil {
+		return fmt.Errorf("unable to create release payload client: %v", err)
+	}
+
+	go mainProcessLoop(ctxWithCancel, stopCh, imageClient.ImageV1(), releasePayloadClient.ReleaseV1alpha1(), o.Namespace, o.Name, o.ProwNamespace, o.Execute, o.Location)
+
+	select {
+	case <-ctx.Done():
+		klog.V(5).Infof("Context has been cancelled")
+	case <-stopCh:
+		klog.V(5).Infof("mainProcessLoop returned")
+	}
+	return nil
+}
+
+func loadClusterConfig(overrides *clientcmd.ConfigOverrides) (*rest.Config, error) {
+	cfg := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(clientcmd.NewDefaultClientConfigLoadingRules(), overrides)
+	clusterConfig, err := cfg.ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load client configuration: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+func mainProcessLoop(ctx context.Context, stopCh chan bool, imageClient imageclient.ImageV1Interface, releasePayloadClient releasev1alpha1.ReleaseV1alpha1Interface, namespace, name, prowNamespace string, execute bool, location string) {
+	ctxWithTimeout, cancelFunction := context.WithTimeout(ctx, time.Duration(15)*time.Second)
+
+	defer func() {
+		cancelFunction()
+		stopCh <- true
+	}()
+
+	var errors []error
+	var payloads []*v1alpha1.ReleasePayload
+	controller := newController(ctxWithTimeout, imageClient, namespace, name, prowNamespace)
+
+	if !execute {
+		klog.Infof("Running in Dry-Run mode.  No objects will be created on cluster.")
+	}
+
+	if len(location) > 0 {
+		payloads = readPayloads(location)
+
+		if execute {
+			errors = createPayloads(releasePayloadClient, payloads)
+		}
+	} else {
+		releaseConfigs, err := controller.processImagestreams()
+		if err != nil {
+			klog.Errorf("Unable to gather release configs: %v", err)
+			return
+		}
+		releases, skipped, problems := controller.processReleaseImagestream()
+		payloads = controller.generatePayloads(releases, releaseConfigs)
+
+		klog.Infof("Successfully generated %d ReleasePayload definitions", len(payloads))
+		klog.Warningf("The following %d tag(s) have problems:\n%s", len(problems), strings.Join(problems, ","))
+		klog.Warningf("The following %d tag(s) cannot be migrated:\n%s", len(skipped), strings.Join(skipped, ","))
+
+		if execute {
+			errors = createPayloads(releasePayloadClient, payloads)
```

Comment:
having now finished my read, I think you need to have your key for the controller be the imagestreamtags.  Your command can list many imagestream tags and compute them all, but you really want your controller to be based on imagestreamtags.  and you want the three controllers I described here https://github.com/openshift/release-controller/pull/348/files#r771476299